mono/metadata/assembly.c

```diff
@@ -1840,26 +1841,77 @@ mono_assembly_load_friends (MonoAssembly* ass)
- * single assembly is missing.  The custom attr we want is from
- * corlib, however, so we need a more robust version that doesn't care
- * about missing attributes.
- */
-#if 1
-	MonoCustomAttrInfo *attrs = mono_custom_attrs_from_assembly_checked (assembly, TRUE, error);
-	return_val_if_nok (error, FALSE);
-	if (!attrs)
+	/*
+	 * This might be called during assembly loading, so do everything using the low-level
+	 * metadata APIs.
+	 */
+
+	image = assembly->image;
+	idx = 1; /* there is only one assembly */
+	idx <<= MONO_CUSTOM_ATTR_BITS;
+	idx |= MONO_CUSTOM_ATTR_ASSEMBLY;
+
+	/* Inlined from mono_custom_attrs_from_index_checked () */
+	ca = &image->tables [MONO_TABLE_CUSTOMATTRIBUTE];
+	i = mono_metadata_custom_attrs_from_index (image, idx);
+	if (!i)
 		return FALSE;
-	MonoClass *ref_asm_class = mono_class_try_get_reference_assembly_class ();
-	g_assert (ref_asm_class != NULL && ref_asm_class != mono_defaults.object_class && !strcmp(ref_asm_class->name, "ReferenceAssemblyAttribute") );
-	gboolean result = mono_custom_attrs_has_attr (attrs, ref_asm_class);
-	mono_custom_attrs_free (attrs);
-	return result;
-#else
+	i --;
+	while (i < ca->rows) {
+		if (mono_metadata_decode_row_col (ca, i, MONO_CUSTOM_ATTR_PARENT) != idx)
+			break;
+		mono_metadata_decode_row (ca, i, cols, MONO_CUSTOM_ATTR_SIZE);
+		i ++;
+		mtoken = cols [MONO_CUSTOM_ATTR_TYPE] >> MONO_CUSTOM_ATTR_TYPE_BITS;
+		if ((cols [MONO_CUSTOM_ATTR_TYPE] & MONO_CUSTOM_ATTR_TYPE_MASK) != MONO_CUSTOM_ATTR_TYPE_MEMBERREF)
+			continue;
+		mtoken |= MONO_TOKEN_MEMBER_REF;
+		{
+			/* method_from_memberref () */
+			guint32 cols[6];
+			guint32 nindex, class_index;
+
+			int idx = mono_metadata_token_index (mtoken);
+
+			mono_metadata_decode_row (&image->tables [MONO_TABLE_MEMBERREF], idx-1, cols, 3);
+			nindex = cols [MONO_MEMBERREF_CLASS] >> MONO_MEMBERREF_PARENT_BITS;
+			class_index = cols [MONO_MEMBERREF_CLASS] & MONO_MEMBERREF_PARENT_MASK;
+			if (class_index == MONO_MEMBERREF_PARENT_TYPEREF) {
+				guint32 type_token = MONO_TOKEN_TYPE_REF | nindex;
+				/* mono_class_from_typeref_checked () */
+				{
+					guint32 cols [MONO_TYPEREF_SIZE];
+					const char *name, *nspace;
+					MonoTableInfo  *t = &image->tables [MONO_TABLE_TYPEREF];
+					guint32 idx;
+
+					mono_metadata_decode_row (t, (type_token&0xffffff)-1, cols, MONO_TYPEREF_SIZE);
+
+					name = mono_metadata_string_heap (image, cols [MONO_TYPEREF_NAME]);
+					nspace = mono_metadata_string_heap (image, cols [MONO_TYPEREF_NAMESPACE]);
+					idx = cols [MONO_TYPEREF_SCOPE] >> MONO_RESOLUTION_SCOPE_BITS;
+
+					if (!strcmp (name, "ReferenceAssemblyAttribute") && !strcmp (nspace, "System.Runtime.CompilerServices")) {
+						if ((cols [MONO_TYPEREF_SCOPE] & MONO_RESOLUTION_SCOPE_MASK) == MONO_RESOLUTION_SCOPE_ASSEMBLYREF) {
+							MonoAssemblyName aname;
+							mono_assembly_get_assemblyref (image, idx - 1, &aname);
+							if (!strcmp (aname.name, "System.Runtime"))
```

Comment:
The assembly actually references:
.custom instance void [System.Runtime]System.Runtime.CompilerServices.ReferenceAssemblyAttribute::.ctor() = ( 01 00 00 00 )